src/Orleans/Threading/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    /// <summary>
+    /// Contains list of cancellation token source corresponding to the tokens
+    /// passed to the related grain activation.
+    /// </summary>
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private static readonly Interner<Guid, GrainCancellationTokenSource> _cancellationTokenSources;
+        private static readonly TimeSpan _cleanupFrequency = TimeSpan.FromMinutes(3);
+        private static readonly int _defaultInternerCollectionSize = 31;
+
+        static CancellationSourcesExtension()
+        {
+            _cancellationTokenSources = new Interner<Guid, GrainCancellationTokenSource>(
```

Comment:
what is the reasoning for using a static field and introducing concurrency, instead of just having a list token sources that you purge from time to time? Maybe it's ok, just wanted to understand it.
Is it so that you don't have to explicitly schedule multiple purges?